File: workbook/ocaml/support/ast_def.ml
(*
   AST ノード定義 — OCaml 版
   文字列ベースの kind 判定をやめ、代数的データ型で表現する。
*)

(* ── 型 ──
   struct はタグ参照（TyStruct "Node"）で表す。フィールドレイアウトは
   Struct_env が管理する（サイズ・整列の計算も Struct_env 側）。 *)

type ty =
  | TyInt
  | TyChar
  | TyVoid
  | TyPtr of ty
  | TyStruct of string

let is_ptr_ty = function TyPtr _ -> true | _ -> false
let is_struct_ty = function TyStruct _ -> true | _ -> false

(* ── 演算子・式・文の AST ── *)

type binop =
  | Add
  | Sub
  | Mul
  | Div
  | Mod
  | Eq
  | Ne
  | Lt
  | Le
  | And
  | Or

type unop =
  | Neg
  | Not
  | Addr
  | Deref
  | PreInc
  | PreDec

(* 前処理後ソース上の UTF-8 バイト半開区間。None は合成ノードを表す。 *)
type source_span = { start_offset : int; end_offset : int }

type expr =
  | Num of { value : int; line : int; span : source_span option }
  | StrLit of { value : string; line : int; span : source_span option }
  | Var of { name : string; line : int; span : source_span option }
  | Assign of { lhs : expr; rhs : expr; line : int; span : source_span option }
  | Cond of { cond : expr; then_ : expr; else_ : expr; line : int; span : source_span option }
  | Binary of { op : binop; lhs : expr; rhs : expr; line : int; span : source_span option }
  | Unary of { op : unop; operand : expr; line : int; span : source_span option }
  | Index of { base : expr; index : expr; line : int; span : source_span option }
  | Member of { base : expr; name : string; is_arrow : bool; line : int; span : source_span option }
  | SizeofType of { ty : ty; line : int; span : source_span option }
  | Call of { name : string; args : expr list; line : int; span : source_span option }

type decl = {
  name : string;
  ty : ty;
  line : int;
  span : source_span option;
}

type stmt =
  | Block of { stmts : stmt list; line : int; span : source_span option }
  | ExprStmt of { expr : expr option; line : int; span : source_span option }
  | Return of { expr : expr option; line : int; span : source_span option }
  | Break of { line : int; span : source_span option }
  | Continue of { line : int; span : source_span option }
  | If of { cond : expr; then_ : stmt; else_ : stmt option; line : int; span : source_span option }
  | While of { cond : expr; body : stmt; line : int; span : source_span option }
  | For of {
      init : expr option;
      cond : expr option;
      step : expr option;
      body : stmt;
      line : int;
      span : source_span option;
    }
  | Decl of decl

type param = {
  name : string option;
  ty : ty;
  line : int;
  span : source_span option;
}

type func = {
  name : string;
  ty : ty;
  params : param list;
  body : stmt;
  line : int;
  span : source_span option;
}

type func_proto = {
  name : string;
  ty : ty;
  params : param list;
  line : int;
  span : source_span option;
}

type top =
  | FuncDef of func
  | FuncProto of func_proto
  | GlobalDecl of decl

type program = top list

let line_of_expr = function
  | Num { line; _ }
  | StrLit { line; _ }
  | Var { line; _ }
  | Assign { line; _ }
  | Cond { line; _ }
  | Binary { line; _ }
  | Unary { line; _ }
  | Index { line; _ }
  | Member { line; _ }
  | SizeofType { line; _ }
  | Call { line; _ } ->
      line

let line_of_stmt = function
  | Block { line; _ }
  | ExprStmt { line; _ }
  | Return { line; _ }
  | Break { line; _ }
  | Continue { line; _ }
  | If { line; _ }
  | While { line; _ }
  | For { line; _ }
  | Decl { line; _ } ->
      line

let span_of_expr = function
  | Num { span; _ }
  | StrLit { span; _ }
  | Var { span; _ }
  | Assign { span; _ }
  | Cond { span; _ }
  | Binary { span; _ }
  | Unary { span; _ }
  | Index { span; _ }
  | Member { span; _ }
  | SizeofType { span; _ }
  | Call { span; _ } ->
      span

let with_expr_span expr span =
  match expr with
  | Num r -> Num { r with span }
  | StrLit r -> StrLit { r with span }
  | Var r -> Var { r with span }
  | Assign r -> Assign { r with span }
  | Cond r -> Cond { r with span }
  | Binary r -> Binary { r with span }
  | Unary r -> Unary { r with span }
  | Index r -> Index { r with span }
  | Member r -> Member { r with span }
  | SizeofType r -> SizeofType { r with span }
  | Call r -> Call { r with span }

File: workbook/scaffold/ast_def.py
"""
AST ノード定義 — 教員提供スキャフォールド

Phase 0–1 全体を通じて使用する。学生はこのファイルを変更しない。
"""

from dataclasses import dataclass, field
from typing import Optional, List

# ---- ノード種別定数 ----

# リテラル
ND_NUM    = 'Num'       # 整数リテラル・文字リテラル  val: int
ND_STR    = 'Str'       # 文字列リテラル              sval: str

# 変数・代入
ND_VAR    = 'Var'       # 変数参照（lvalue / rvalue）  name: str
ND_ASSIGN = 'Assign'    # =                            lhs, rhs

# 二項算術
ND_ADD = 'Add'          # +
ND_SUB = 'Sub'          # -
ND_MUL = 'Mul'          # *
ND_DIV = 'Div'          # /
ND_MOD = 'Mod'          # %

# 単項
ND_NEG    = 'Neg'       # 単項 -         operand
ND_NOT    = 'Not'       # !              operand
ND_ADDR   = 'Addr'      # & アドレス取得  operand
ND_DEREF  = 'Deref'     # * 間接参照      operand
ND_PREINC = 'PreInc'    # 前置 ++        operand
ND_PREDEC = 'PreDec'    # 前置 --        operand

# 比較（> / >= は lhs・rhs を swap して LT / LE に正規化）
ND_EQ = 'Eq'            # ==
ND_NE = 'Ne'            # !=
ND_LT = 'Lt'            # <   （a > b は Node(LT, lhs=b, rhs=a) で表現）
ND_LE = 'Le'            # <=  （a >= b は Node(LE, lhs=b, rhs=a) で表現）

# 論理
ND_AND = 'And'          # &&
ND_OR  = 'Or'           # ||

# 条件（三項）
ND_COND = 'Cond'        # a ? b : c      cond, then, else_

# ポインタ・複合
ND_INDEX  = 'Index'     # p[i]           lhs=ポインタ, rhs=添字
ND_MEMBER = 'Member'    # x.f / p->f     operand, name, is_arrow: bool

# sizeof（型名形式のみ）
ND_SIZEOF_TYPE = 'SizeofType'   # sizeof(type)  ty_str

# 関数呼び出し
ND_CALL = 'Call'        # f(args)        name, args: List[Node]

# 文
ND_BLOCK    = 'Block'       # { stmts }           stmts: List[Node]
ND_EXPRSTMT = 'ExprStmt'    # expr;               operand (None = 空文)
ND_RETURN   = 'Return'      # return [expr];      operand (None = void)
ND_BREAK    = 'Break'       # break;
ND_CONTINUE = 'Continue'    # continue;
ND_IF       = 'If'          # if/else             cond, then, else_
ND_WHILE    = 'While'       # while               cond, body
ND_FOR      = 'For'         # for                 init, cond, step, body
ND_DECL     = 'Decl'        # 変数宣言（初期化子なし）  name, ty_str

# トップレベル
ND_FUNCDEF   = 'FuncDef'    # 関数定義  name, ty_str, params, body
ND_FUNCPROTO = 'FuncProto'  # 関数宣言  name, ty_str, params


@dataclass
class Node:
    kind: str

    # 二項演算 / if・三項の各枝
    lhs:   Optional['Node'] = None
    rhs:   Optional['Node'] = None
    cond:  Optional['Node'] = None
    then:  Optional['Node'] = None
    else_: Optional['Node'] = None

    # ループ
    init: Optional['Node'] = None
    step: Optional['Node'] = None
    body: Optional['Node'] = None

    # 単項演算・return のオペランド
    operand: Optional['Node'] = None

    # リスト形フィールド
    stmts:  List['Node'] = field(default_factory=list)   # Block の本体
    args:   List['Node'] = field(default_factory=list)   # 関数呼び出し引数
    params: List['Node'] = field(default_factory=list)   # 関数パラメータ (Decl)

    # 値・名前
    val:  int = 0    # ND_NUM
    sval: str = ''   # ND_STR（文字列内容）
    name: str = ''   # ND_VAR / ND_CALL / ND_MEMBER / ND_FUNCDEF / ND_DECL

    # 型情報（文字列表現: 'int', 'char', 'int*', 'struct Node*', ...）
    # コード生成器が付与してもよいし、Parser が付与してもよい
    ty_str: str = ''

    # Member: -> か . か
    is_arrow: bool = False

    # デバッグ用（字句トークンの行番号）
    line: int = 0

    def __repr__(self):
        return f"Node({self.kind!r}, name={self.name!r}, val={self.val})"
